### PR TITLE
Laboratory work #3, Marina Mokina - 22FPL1

### PR DESCRIPTION
**Pull request by mmarina2004**

### CODE REVIEW

File: lab_3_generate_by_ngrams/main.py [hunk 1]
```diff
@@ -23,6 +23,8 @@ def __init__(self, end_of_word_token: str) -> None:
         Args:
             end_of_word_token (str): A token denoting word boundary
         """
+        self._end_of_word_token = end_of_word_token
+        self._storage = {'_': 0}
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:27]: why do you use this constant?

File: lab_3_generate_by_ngrams/main.py [hunk 2]
```diff
@@ -41,6 +43,19 @@ def _tokenize(self, text: str) -> Optional[tuple[str, ...]]:
         In case of corrupt input arguments, None is returned.
         In case any of methods used return None, None is returned.
         """
+        if not isinstance(text, str) or text == '':
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:46]: `text == ''` -> `not text`

File: lab_3_generate_by_ngrams/main.py [hunk 3]
```diff
@@ -41,6 +43,19 @@ def _tokenize(self, text: str) -> Optional[tuple[str, ...]]:
...
+            return None
+
+        list_text = text.split()
+        str_text = self._end_of_word_token.join(list_text)
+        new_str = ''
+        for token in str_text:
+            if token.isalpha() or token == '_':
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:53]: again strange constant

File: lab_3_generate_by_ngrams/main.py [hunk 4]
```diff
@@ -41,6 +43,19 @@ def _tokenize(self, text: str) -> Optional[tuple[str, ...]]:
...
+                new_str += token
+        new_str += '_'
+        if '__' in new_str:
+            copy_str = new_str.replace('__', '_')
+            return tuple([token.lower() for token in copy_str])
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:58]: Convert sting to lower case  and than just list. Without any loops

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_3_generate_by_ngrams/main.py (hunk 1 above)
Comment (mmarina2004) [reply lab_3_generate_by_ngrams/main.py]: changed it

File: lab_3_generate_by_ngrams/main.py (hunk 2 above)
Comment (mmarina2004) [reply lab_3_generate_by_ngrams/main.py]: done

File: lab_3_generate_by_ngrams/main.py (hunk 4 above)
Comment (mmarina2004) [reply lab_3_generate_by_ngrams/main.py]: done

File: lab_3_generate_by_ngrams/main.py [hunk 5]
```diff
@@ -76,6 +98,12 @@ def get_token(self, element_id: int) -> Optional[str]:
 
         In case of corrupt input arguments or arguments not included in storage, None is returned
         """
+        if not isinstance(element_id, int) or element_id not in self._storage.values():
+            return None
+
+        for key in self._storage:
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:104]: try to use `.items()` to avoid using the `[]` operator

File: lab_3_generate_by_ngrams/main.py [hunk 6]
```diff
@@ -104,6 +147,10 @@ def _put(self, element: str) -> None:
         In case of corrupt input arguments or invalid argument length,
         an element is not added to storage
         """
+        if not isinstance(element, str) or len(element) != 1:
+            return None
+        if element not in self._storage:
+            self._storage[element] = len(self._storage)
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:153]: looks like you get pylint error in this place

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]
Good start, try to get 8 and 10!

File: lab_3_generate_by_ngrams/main.py (hunk 5 above)
Comment (mmarina2004) [reply lab_3_generate_by_ngrams/main.py]: done

File: lab_3_generate_by_ngrams/main.py (hunk 6 above)
Comment (mmarina2004) [reply lab_3_generate_by_ngrams/main.py:154]: fixed

File: lab_3_generate_by_ngrams/main.py [hunk 7]
```diff
@@ -41,6 +44,21 @@ def _tokenize(self, text: str) -> Optional[tuple[str, ...]]:
         In case of corrupt input arguments, None is returned.
         In case any of methods used return None, None is returned.
         """
+        if not isinstance(text, str):
+            return None
+
+        tokens = []
+        list_text = text.lower().split()
+        for element in list_text:
+            word = [token for token in element if token.isalpha()]
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:53]: `filter` + `extend`

File: lab_3_generate_by_ngrams/main.py [hunk 8]
```diff
@@ -76,6 +99,12 @@ def get_token(self, element_id: int) -> Optional[str]:
 
         In case of corrupt input arguments or arguments not included in storage, None is returned
         """
+        if not isinstance(element_id, int) or element_id not in self._storage.values():
+            return None
+
+        for key, value in self._storage.items():
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:105]: filter

File: lab_3_generate_by_ngrams/main.py [hunk 9]
```diff
@@ -104,6 +148,11 @@ def _put(self, element: str) -> None:
         In case of corrupt input arguments or invalid argument length,
         an element is not added to storage
         """
+        if not isinstance(element, str) or len(element) != 1:
+            return None
+        if element not in self._storage:
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:153]: reverse the condition and do early return

File: lab_3_generate_by_ngrams/main.py [hunk 10]
```diff
@@ -129,6 +188,12 @@ def fill_from_ngrams(self, content: dict) -> None:
         Args:
             content (dict): ngrams from external JSON
         """
+        if not isinstance(content, dict) or not content:
+            return None
+        for key in content['freq']:
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:193]: toooo dificult

Comment (artyomtugaryov) [reply lab_3_generate_by_ngrams/main.py:193]: eliminate one for

File: lab_3_generate_by_ngrams/main.py [hunk 11]
```diff
@@ -159,6 +236,18 @@ def _postprocess_decoded_text(self, decoded_corpus: tuple[str, ...]) -> Optional
...
+
+        decoded_text = decoded_corpus[0].upper()
+        for token in decoded_corpus[1:-1]:
+            if token == self._end_of_word_token:
+                decoded_text += ' '
+            else:
+                decoded_text += token
+        if decoded_corpus[-1] != self._end_of_word_token:
+            decoded_text += decoded_corpus[-1]
+        return decoded_text + '.'
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:250]: fstring

File: lab_3_generate_by_ngrams/main.py [hunk 12]
```diff
@@ -159,6 +236,18 @@ def _postprocess_decoded_text(self, decoded_corpus: tuple[str, ...]) -> Optional
 
         In case of corrupt input arguments, None is returned
         """
+        if not isinstance(decoded_corpus, tuple) or len(decoded_corpus) == 0:
+            return None
+
+        decoded_text = decoded_corpus[0].upper()
+        for token in decoded_corpus[1:-1]:
+            if token == self._end_of_word_token:
+                decoded_text += ' '
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:245]: fstring

File: lab_3_generate_by_ngrams/main.py [hunk 13]
```diff
@@ -208,6 +305,21 @@ def build(self) -> int:
...
+
+        n_grams = self._extract_n_grams(self._encoded_corpus)
+        if not n_grams:
+            return 1
+
+        for n_gram in set(n_grams):
+            if not isinstance(n_gram, tuple):
+                return 1
+            absolute_frequency = n_grams.count(n_gram)
+            with_same_beginning = len([id for id in n_grams if
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:319]: count

File: lab_3_generate_by_ngrams/main.py [hunk 14]
```diff
@@ -270,6 +403,27 @@ def run(self, seq_len: int, prompt: str) -> Optional[str]:
...
+        n_gram_size = self._model.get_n_gram_size()
+        encoded = self._text_processor.encode(prompt)
+        if not encoded or not n_gram_size:
+            return None
+
+        while seq_len > 0:
+            candidates = self._model.generate_next_token(encoded[-n_gram_size + 1:])
+            if not candidates:
+                break
+            best_candidate = [letter for letter, freq in candidates.items() if freq == max(candidates.values())]
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:418]: filter and get `max(candidates.values())` only once

File: lab_3_generate_by_ngrams/main.py [hunk 15]
```diff
@@ -309,6 +465,17 @@ def get_next_token(self, sequence: tuple[int, ...]) -> Optional[list[tuple[int,
...
+        if not isinstance(sequence, tuple) or len(sequence) == 0:
+            return None
+
+        tokens_dict = self._model.generate_next_token(sequence)
+        if tokens_dict is None:
+            return None
+        if not tokens_dict:
+            return []
+
+        return sorted([(token, float(probability)) for token, probability in tokens_dict.items()],
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:477]: why convert to float only once?

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]
Only 6?

File: lab_3_generate_by_ngrams/main.py (hunk 8 above)
Comment (mmarina2004) [reply lab_3_generate_by_ngrams/main.py]: done

File: lab_3_generate_by_ngrams/main.py (hunk 7 above)
Comment (artyomtugaryov) [reply lab_3_generate_by_ngrams/main.py:53]: extend is already used

File: lab_3_generate_by_ngrams/main.py [hunk 16]
```diff
@@ -236,6 +359,15 @@ def _extract_n_grams(
 
         In case of corrupt input arguments, None is returned
         """
+        if not isinstance(encoded_corpus, tuple) or len(encoded_corpus) == 0:
+            return None
+
+        n_grams = []
+        list_encoded_corpus = list(encoded_corpus)
+        for index in range(len(encoded_corpus) + 1 - self._n_gram_size):
+            n_grams.append(tuple(list_encoded_corpus[index: index + self._n_gram_size]))
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:368]: please, save the tuple to a variable. It is dificult to read

File: lab_3_generate_by_ngrams/main.py [hunk 17]
```diff
@@ -331,6 +496,27 @@ def continue_sequence(
...
+                len(next_tokens) > self._beam_width):
+            return None
+
+        new_sequence_candidates = {}
+        for key, value in sequence_candidates.items():
+            if key != sequence:
+                probability = value
+                for token, token_probability in next_tokens:
+                    new_sequence = key + (token,)
+                    new_probability = probability + (-1) * math.log(token_probability)
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:513]: why do you need the multiplication? can you just substitute log?

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_3_generate_by_ngrams/main.py (hunk 7 above)
Comment (mmarina2004) [reply lab_3_generate_by_ngrams/main.py]: done

File: lab_3_generate_by_ngrams/main.py (hunk 11 above)
Comment (mmarina2004) [reply lab_3_generate_by_ngrams/main.py]: done

File: lab_3_generate_by_ngrams/main.py (hunk 12 above)
Comment (mmarina2004) [reply lab_3_generate_by_ngrams/main.py]: done

File: lab_3_generate_by_ngrams/main.py (hunk 16 above)
Comment (mmarina2004) [reply lab_3_generate_by_ngrams/main.py]: done

File: lab_3_generate_by_ngrams/main.py (hunk 17 above)
Comment (mmarina2004) [reply lab_3_generate_by_ngrams/main.py]: changed it

File: lab_3_generate_by_ngrams/main.py (hunk 15 above)
Comment (mmarina2004) [reply lab_3_generate_by_ngrams/main.py:473]: I only return this tuple (token, float(probability))

File: lab_3_generate_by_ngrams/main.py (hunk 14 above)
Comment (mmarina2004) [reply lab_3_generate_by_ngrams/main.py]: It didn't work out

File: lab_3_generate_by_ngrams/main.py (hunk 13 above)
Comment (mmarina2004) [reply lab_3_generate_by_ngrams/main.py:317]: It didn't work out

File: lab_3_generate_by_ngrams/main.py (hunk 10 above)
Comment (mmarina2004) [reply lab_3_generate_by_ngrams/main.py]: It didn't work out

**Pull request closed without merging**